jfix-zookeeper/src/main/kotlin/ru/fix/zookeeper/instance/registry/ServiceInstanceIdRegistry.kt

```diff
@@ -0,0 +1,196 @@
+import ru.fix.zookeeper.utils.Marshaller
+import ru.fix.zookeeper.utils.ZkTreePrinter
+import java.time.Duration
+import java.time.Instant
+
+/**
+ * This class provides functionality of instance id management.
+ * When this class instantiated, it generates new instance id for this zookeeper client and registers this id in znode.
+ * When zookeeper client lost connection, instance will get previous instance after reconnection.
+ * If new client connected, if disconnect timeout expired and client's config.host and config.port already contains in znodes,
+ * it is seen as instance restart, and in this case instance get instance id, already contained in zk.
+ * Service registration guarantees uniqueness of instance id in case of parallel startup on different JVMs.
+ */
+class ServiceInstanceIdRegistry(
+        private val curatorFramework: CuratorFramework,
+        private val instanceIdGenerator: InstanceIdGenerator,
+        private val config: ServiceInstanceIdRegistryConfig,
+        profiler: Profiler
+) : AutoCloseable {
+    lateinit var lockIdentity: LockIdentity
+        private set
+    lateinit var instanceIdPath: String
+        private set
+    lateinit var instanceId: String
+        private set
+
+    companion object {
+        private val logger = LoggerFactory.getLogger(ServiceInstanceIdRegistry::class.java)
+    }
+
+    private val lockManager: PersistentExpiringLockManager = PersistentExpiringLockManager(
+            curatorFramework,
+            DynamicProperty.of(
+                    PersistentExpiringLockManagerConfig(
+                            reservationPeriod = config.disconnectTimeout.dividedBy(2),
+                            expirationPeriod = config.disconnectTimeout.dividedBy(3),
+                            lockProlongationInterval = config.disconnectTimeout.dividedBy(4),
+                            acquiringTimeout = Duration.ofSeconds(1)
+                    )
+            ),
+            profiler
+    )
+
+    init {
+        curatorFramework
+                .connectionStateListenable
+                .addListener(
+                        ConnectionStateListener { _, newState ->
+                            when (newState) {
+                                ConnectionState.RECONNECTED -> reconnect()
+                                else -> Unit
+                            }
+                        })
+        initServiceRegistrationPath()
+        initInstanceId()
+    }
+
+    private fun initServiceRegistrationPath() {
+        try {
+            curatorFramework.create()
+                    .creatingParentsIfNeeded()
+                    .forPath(config.serviceRegistrationPath, byteArrayOf())
+        } catch (e: KeeperException.NodeExistsException) {
+            logger.debug("Node ${config.serviceRegistrationPath} is already initialized", e)
+        } catch (e: Exception) {
+            logger.error("Illegal state when create path: ${config.serviceRegistrationPath}", e)
```

Comment:
Why you are hiding Exception and allowing user to thing that ServiceInstanceRegistry correctly initialized, while it is not?

Reply:
throw on error